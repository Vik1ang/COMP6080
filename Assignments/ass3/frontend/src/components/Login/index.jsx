import React from 'react';
import {
  Avatar,
  Box,
  Button,
  Checkbox,
  Container,
  CssBaseline,
  FormControlLabel,
  Grid,
  TextField,
  ThemeProvider,
  Typography,
} from '@mui/material';
import { Link, useHistory } from "react-router-dom";
import localforage from "localforage";
import {Login as LoginIcon} from '@mui/icons-material'
import { requestLogin } from '../../services/request';
import theme from "../../theme";
import msgPopup from "../../utils/msgPopup";
import { sleep } from "../../utils/utils";

export default function Login() {

  const history = useHistory();

  const handleLogin = event => {
    event.preventDefault();
    const data = new FormData(event.currentTarget);

    const payload = {
      email: data.get('email'),
      password: data.get('password'),
    };

    requestLogin(payload)
      .then(async response => {
        await localforage.setItem('token', response.token);
        await localforage.setItem('user', payload.email);
        msgPopup('Login Successfully', 'success');
        await sleep(2000);
        return history.push('/');
      })
      .catch(err => ({ ...err }));
  };
  return (
    <ThemeProvider theme={theme}>
      <Container component='main' maxWidth='xs'>
        <CssBaseline />
        <Box
          sx={{
            marginTop: 8,
            display: 'flex',
            flexDirection: 'column',
            alignItems: 'center',
          }}>
          <Avatar sx={{ m: 1, bgcolor: 'secondary.main' }}>
            <LoginIcon />
          </Avatar>
          <Typography component='h1' variant='h5'>
            Log In
          </Typography>
          <Box component='form' noValidate onSubmit={handleLogin} sx={{ mt: 3 }}>
            <TextField
              margin='normal'
              required
              fullWidth
              id='email'
              label='Email Address'
              name='email'
              autoComplete='email'
              autoFocus
            />
            <TextField
              margin='normal'
              required
              fullWidth
              name='password'
              label='Password'
              type='password'
              id='password'
              autoComplete='current-password'
            />
            <FormControlLabel control={<Checkbox value='remember' color='primary' />} label='Remember me' />
            <Button type='submit' fullWidth variant='contained' sx={{ mt: 3, mb: 2 }}>
              Sign in
            </Button>
            <Grid container justifyContent='flex-end'>
              <Grid item>
                {/* eslint-disable-next-line jsx-a11y/anchor-is-valid */}
                <Link to='/register' variant='body2'>
                  {'Don\'t have an account? Sign Up'}
                </Link>
              </Grid>
            </Grid>
          </Box>
        </Box>
      </Container>
    </ThemeProvider>
  );
}
