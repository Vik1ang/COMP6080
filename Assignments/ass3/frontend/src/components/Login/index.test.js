// eslint-disable-next-line no-unused-vars
import {shallow} from 'enzyme';
// eslint-disable-next-line no-unused-vars
import React from 'react';
import renderer from 'react-test-renderer';
// eslint-disable-next-line no-unused-vars
import toJson from 'enzyme-to-json';
// eslint-disable-next-line no-unused-vars
import { act } from 'react-dom/test-utils';
import { BrowserRouter as Router, Route } from 'react-router-dom';
import { cleanup, fireEvent, render, waitFor } from '@testing-library/react';
import Login from './index';
import {
    Typography,
    TextField,
    Button,
    FormControlLabel,
    Grid,
  } from '@mui/material';

const props_one = {
  email: 'selinajiang321@gmail.com',
  password: "123",
};


afterEach(cleanup);

describe('Login', () => {
  const noop = () => {};
  it('render Login correctly', () => {
    const login = shallow(<Login/>);
    expect(toJson(login)).toMatchSnapshot();
  });

  it('render Login Button correctly', () => {
    const wrapper = shallow(<Login />);
    // const { email } = props;
    const logInDom = (
        <Typography component='h1' variant='h5'>
        Log In
      </Typography>
    );
    expect(wrapper.contains(logInDom)).toEqual(true);
  });

  test('TextField component should exists.', () => {
    expect(TextField).toBeDefined();
  });

  it('renders a button with submit button type if nothing is provided', () => {
    const button = renderer
      .create(<Button onClick={noop} type='submit' />)
      .toJSON();
    expect(button).toMatchSnapshot();
  });


  it('has one button component', () => {
    const button = shallow(<Login />);
    expect(button.find(Button)).toHaveLength(1);
  });

  it('has two TextField components', () => {
    const textField = shallow(<Login />);
    expect(textField.find(TextField)).toHaveLength(2);
  });

  it('has a link inside a grid box to register', () => {
    const loginPage = shallow(<Login />);
    expect(loginPage.find(Grid)).toHaveLength(2);
  });

  it('has a checkbox to readme', () => {
    const loginPage = shallow(<Login />);
    expect(loginPage.find(FormControlLabel)).toHaveLength(1);
  });

  it('has a link to register', () => {
    const { getByText } = render(
      <Router>
        <Route>
          <Login />
        </Route>
      </Router>,
    );
    act (() => {fireEvent.click(getByText('Don\'t have an account? Sign Up'))});
    expect(window.location.href).toEqual(`http://localhost/register`);
  });


});
