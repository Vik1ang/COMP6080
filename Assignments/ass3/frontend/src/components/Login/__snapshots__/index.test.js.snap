// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login render Login correctly 1`] = `
<ThemeProvider
  theme={
    Object {
      "breakpoints": Object {
        "between": [Function],
        "down": [Function],
        "keys": Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ],
        "only": [Function],
        "unit": "px",
        "up": [Function],
        "values": Object {
          "lg": 1200,
          "md": 900,
          "sm": 600,
          "xl": 1536,
          "xs": 0,
        },
      },
      "components": Object {},
      "direction": "ltr",
      "mixins": Object {
        "toolbar": Object {
          "@media (min-width:0px) and (orientation: landscape)": Object {
            "minHeight": 48,
          },
          "@media (min-width:600px)": Object {
            "minHeight": 64,
          },
          "minHeight": 56,
        },
      },
      "palette": Object {
        "action": Object {
          "activatedOpacity": 0.12,
          "active": "rgba(0, 0, 0, 0.54)",
          "disabled": "rgba(0, 0, 0, 0.26)",
          "disabledBackground": "rgba(0, 0, 0, 0.12)",
          "disabledOpacity": 0.38,
          "focus": "rgba(0, 0, 0, 0.12)",
          "focusOpacity": 0.12,
          "hover": "rgba(0, 0, 0, 0.04)",
          "hoverOpacity": 0.04,
          "selected": "rgba(0, 0, 0, 0.08)",
          "selectedOpacity": 0.08,
        },
        "augmentColor": [Function],
        "background": Object {
          "default": "#fff",
          "paper": "#fff",
        },
        "common": Object {
          "black": "#000",
          "white": "#fff",
        },
        "contrastThreshold": 3,
        "divider": "rgba(0, 0, 0, 0.12)",
        "error": Object {
          "contrastText": "#fff",
          "dark": "rgb(178, 16, 47)",
          "light": "rgb(255, 69, 105)",
          "main": "#ff1744",
        },
        "getContrastText": [Function],
        "grey": Object {
          "100": "#f5f5f5",
          "200": "#eeeeee",
          "300": "#e0e0e0",
          "400": "#bdbdbd",
          "50": "#fafafa",
          "500": "#9e9e9e",
          "600": "#757575",
          "700": "#616161",
          "800": "#424242",
          "900": "#212121",
          "A100": "#f5f5f5",
          "A200": "#eeeeee",
          "A400": "#bdbdbd",
          "A700": "#616161",
        },
        "info": Object {
          "contrastText": "#fff",
          "dark": "#01579b",
          "light": "#03a9f4",
          "main": "#0288d1",
        },
        "mode": "light",
        "primary": Object {
          "contrastText": "#fff",
          "dark": "rgb(163, 21, 69)",
          "light": "rgb(237, 75, 130)",
          "main": "#e91e63",
        },
        "secondary": Object {
          "contrastText": "rgba(0, 0, 0, 0.87)",
          "dark": "rgb(53, 122, 56)",
          "light": "rgb(111, 191, 115)",
          "main": "#4caf50",
        },
        "success": Object {
          "contrastText": "#fff",
          "dark": "#1b5e20",
          "light": "#4caf50",
          "main": "#2e7d32",
        },
        "text": Object {
          "disabled": "rgba(0, 0, 0, 0.38)",
          "primary": "rgba(0, 0, 0, 0.87)",
          "secondary": "rgba(0, 0, 0, 0.6)",
        },
        "tonalOffset": 0.2,
        "warning": Object {
          "contrastText": "#fff",
          "dark": "#e65100",
          "light": "#ff9800",
          "main": "#ed6c02",
        },
      },
      "shadows": Array [
        "none",
        "0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12)",
        "0px 3px 1px -2px rgba(0,0,0,0.2),0px 2px 2px 0px rgba(0,0,0,0.14),0px 1px 5px 0px rgba(0,0,0,0.12)",
        "0px 3px 3px -2px rgba(0,0,0,0.2),0px 3px 4px 0px rgba(0,0,0,0.14),0px 1px 8px 0px rgba(0,0,0,0.12)",
        "0px 2px 4px -1px rgba(0,0,0,0.2),0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12)",
        "0px 3px 5px -1px rgba(0,0,0,0.2),0px 5px 8px 0px rgba(0,0,0,0.14),0px 1px 14px 0px rgba(0,0,0,0.12)",
        "0px 3px 5px -1px rgba(0,0,0,0.2),0px 6px 10px 0px rgba(0,0,0,0.14),0px 1px 18px 0px rgba(0,0,0,0.12)",
        "0px 4px 5px -2px rgba(0,0,0,0.2),0px 7px 10px 1px rgba(0,0,0,0.14),0px 2px 16px 1px rgba(0,0,0,0.12)",
        "0px 5px 5px -3px rgba(0,0,0,0.2),0px 8px 10px 1px rgba(0,0,0,0.14),0px 3px 14px 2px rgba(0,0,0,0.12)",
        "0px 5px 6px -3px rgba(0,0,0,0.2),0px 9px 12px 1px rgba(0,0,0,0.14),0px 3px 16px 2px rgba(0,0,0,0.12)",
        "0px 6px 6px -3px rgba(0,0,0,0.2),0px 10px 14px 1px rgba(0,0,0,0.14),0px 4px 18px 3px rgba(0,0,0,0.12)",
        "0px 6px 7px -4px rgba(0,0,0,0.2),0px 11px 15px 1px rgba(0,0,0,0.14),0px 4px 20px 3px rgba(0,0,0,0.12)",
        "0px 7px 8px -4px rgba(0,0,0,0.2),0px 12px 17px 2px rgba(0,0,0,0.14),0px 5px 22px 4px rgba(0,0,0,0.12)",
        "0px 7px 8px -4px rgba(0,0,0,0.2),0px 13px 19px 2px rgba(0,0,0,0.14),0px 5px 24px 4px rgba(0,0,0,0.12)",
        "0px 7px 9px -4px rgba(0,0,0,0.2),0px 14px 21px 2px rgba(0,0,0,0.14),0px 5px 26px 4px rgba(0,0,0,0.12)",
        "0px 8px 9px -5px rgba(0,0,0,0.2),0px 15px 22px 2px rgba(0,0,0,0.14),0px 6px 28px 5px rgba(0,0,0,0.12)",
        "0px 8px 10px -5px rgba(0,0,0,0.2),0px 16px 24px 2px rgba(0,0,0,0.14),0px 6px 30px 5px rgba(0,0,0,0.12)",
        "0px 8px 11px -5px rgba(0,0,0,0.2),0px 17px 26px 2px rgba(0,0,0,0.14),0px 6px 32px 5px rgba(0,0,0,0.12)",
        "0px 9px 11px -5px rgba(0,0,0,0.2),0px 18px 28px 2px rgba(0,0,0,0.14),0px 7px 34px 6px rgba(0,0,0,0.12)",
        "0px 9px 12px -6px rgba(0,0,0,0.2),0px 19px 29px 2px rgba(0,0,0,0.14),0px 7px 36px 6px rgba(0,0,0,0.12)",
        "0px 10px 13px -6px rgba(0,0,0,0.2),0px 20px 31px 3px rgba(0,0,0,0.14),0px 8px 38px 7px rgba(0,0,0,0.12)",
        "0px 10px 13px -6px rgba(0,0,0,0.2),0px 21px 33px 3px rgba(0,0,0,0.14),0px 8px 40px 7px rgba(0,0,0,0.12)",
        "0px 10px 14px -6px rgba(0,0,0,0.2),0px 22px 35px 3px rgba(0,0,0,0.14),0px 8px 42px 7px rgba(0,0,0,0.12)",
        "0px 11px 14px -7px rgba(0,0,0,0.2),0px 23px 36px 3px rgba(0,0,0,0.14),0px 9px 44px 8px rgba(0,0,0,0.12)",
        "0px 11px 15px -7px rgba(0,0,0,0.2),0px 24px 38px 3px rgba(0,0,0,0.14),0px 9px 46px 8px rgba(0,0,0,0.12)",
      ],
      "shape": Object {
        "borderRadius": 4,
      },
      "spacing": [Function],
      "transitions": Object {
        "create": [Function],
        "duration": Object {
          "complex": 375,
          "enteringScreen": 225,
          "leavingScreen": 195,
          "short": 250,
          "shorter": 200,
          "shortest": 150,
          "standard": 300,
        },
        "easing": Object {
          "easeIn": "cubic-bezier(0.4, 0, 1, 1)",
          "easeInOut": "cubic-bezier(0.4, 0, 0.2, 1)",
          "easeOut": "cubic-bezier(0.0, 0, 0.2, 1)",
          "sharp": "cubic-bezier(0.4, 0, 0.6, 1)",
        },
        "getAutoHeightDuration": [Function],
      },
      "typography": Object {
        "body1": Object {
          "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
          "fontSize": "1rem",
          "fontWeight": 400,
          "letterSpacing": "0.00938em",
          "lineHeight": 1.5,
        },
        "body2": Object {
          "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
          "fontSize": "0.875rem",
          "fontWeight": 400,
          "letterSpacing": "0.01071em",
          "lineHeight": 1.43,
        },
        "button": Object {
          "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
          "fontSize": "0.875rem",
          "fontWeight": 500,
          "letterSpacing": "0.02857em",
          "lineHeight": 1.75,
          "textTransform": "uppercase",
        },
        "caption": Object {
          "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
          "fontSize": "0.75rem",
          "fontWeight": 400,
          "letterSpacing": "0.03333em",
          "lineHeight": 1.66,
        },
        "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
        "fontSize": 14,
        "fontWeightBold": 700,
        "fontWeightLight": 300,
        "fontWeightMedium": 500,
        "fontWeightRegular": 400,
        "h1": Object {
          "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
          "fontSize": "6rem",
          "fontWeight": 300,
          "letterSpacing": "-0.01562em",
          "lineHeight": 1.167,
        },
        "h2": Object {
          "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
          "fontSize": "3.75rem",
          "fontWeight": 300,
          "letterSpacing": "-0.00833em",
          "lineHeight": 1.2,
        },
        "h3": Object {
          "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
          "fontSize": "3rem",
          "fontWeight": 400,
          "letterSpacing": "0em",
          "lineHeight": 1.167,
        },
        "h4": Object {
          "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
          "fontSize": "2.125rem",
          "fontWeight": 400,
          "letterSpacing": "0.00735em",
          "lineHeight": 1.235,
        },
        "h5": Object {
          "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
          "fontSize": "1.5rem",
          "fontWeight": 400,
          "letterSpacing": "0em",
          "lineHeight": 1.334,
        },
        "h6": Object {
          "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
          "fontSize": "1.25rem",
          "fontWeight": 500,
          "letterSpacing": "0.0075em",
          "lineHeight": 1.6,
        },
        "htmlFontSize": 16,
        "overline": Object {
          "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
          "fontSize": "0.75rem",
          "fontWeight": 400,
          "letterSpacing": "0.08333em",
          "lineHeight": 2.66,
          "textTransform": "uppercase",
        },
        "pxToRem": [Function],
        "subtitle1": Object {
          "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
          "fontSize": "1rem",
          "fontWeight": 400,
          "letterSpacing": "0.00938em",
          "lineHeight": 1.75,
        },
        "subtitle2": Object {
          "fontFamily": "\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif",
          "fontSize": "0.875rem",
          "fontWeight": 500,
          "letterSpacing": "0.00714em",
          "lineHeight": 1.57,
        },
      },
      "zIndex": Object {
        "appBar": 1100,
        "drawer": 1200,
        "mobileStepper": 1000,
        "modal": 1300,
        "snackbar": 1400,
        "speedDial": 1050,
        "tooltip": 1500,
      },
    }
  }
>
  <ForwardRef(Container)
    component="main"
    maxWidth="xs"
  >
    <CssBaseline />
    <ForwardRef(Box)
      sx={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "column",
          "marginTop": 8,
        }
      }
    >
      <ForwardRef(Avatar)
        sx={
          Object {
            "bgcolor": "secondary.main",
            "m": 1,
          }
        }
      >
        <Memo(ForwardRef(LoginIcon)) />
      </ForwardRef(Avatar)>
      <ForwardRef(Typography)
        component="h1"
        variant="h5"
      >
        Log In
      </ForwardRef(Typography)>
      <ForwardRef(Box)
        component="form"
        noValidate={true}
        onSubmit={[Function]}
        sx={
          Object {
            "mt": 3,
          }
        }
      >
        <ForwardRef(TextField)
          autoComplete="email"
          autoFocus={true}
          fullWidth={true}
          id="email"
          label="Email Address"
          margin="normal"
          name="email"
          required={true}
        />
        <ForwardRef(TextField)
          autoComplete="current-password"
          fullWidth={true}
          id="password"
          label="Password"
          margin="normal"
          name="password"
          required={true}
          type="password"
        />
        <ForwardRef(FormControlLabel)
          control={
            <ForwardRef(Checkbox)
              color="primary"
              value="remember"
            />
          }
          label="Remember me"
        />
        <ForwardRef(Button)
          fullWidth={true}
          sx={
            Object {
              "mb": 2,
              "mt": 3,
            }
          }
          type="submit"
          variant="contained"
        >
          Sign in
        </ForwardRef(Button)>
        <ForwardRef(Grid)
          container={true}
          justifyContent="flex-end"
        >
          <ForwardRef(Grid)
            item={true}
          >
            <Link
              to="/register"
              variant="body2"
            >
              Don't have an account? Sign Up
            </Link>
          </ForwardRef(Grid)>
        </ForwardRef(Grid)>
      </ForwardRef(Box)>
    </ForwardRef(Box)>
  </ForwardRef(Container)>
</ThemeProvider>
`;

exports[`Login renders a button with submit button type if nothing is provided 1`] = `
<button
  className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="submit"
/>
`;
